### DISCUSSION TITLE
How to use Redis as a cache instead of Caffeine?

### DISCUSSION THREAD

**Discussion by dewan-ahmed** [quarkusio/quarkus#24517] [Q&A] [answered]
I'm trying to build an app using Quarkus and Redis as the cache. I'm following two guides:
1. [Application Data Caching guide](https://quarkus.io/guides/cache) which uses Caffeine as the cache, and
2. [Redis client guide](https://quarkus.io/guides/redis) which uses Redis increment/decrement functionality.
 
Could someone please point out how I can use Redis as the cache instead of the underlying Caffeine caching mechanism?

[My work-in-progress code](https://raw.githubusercontent.com/dewan-ahmed/quarkus-redis-cache/main/src/main/java/org/acme/cache/GithubStarsService.java)

This is my assumption of what needs to happen (I don't know the 'how'):

1. I'll have to have a Redis server - either locally via Docker or some managed Redis service. I'll need to enter the Redis server details in the application.properties file.
2. I'll need to use a Redis client like Redisson and add a dependency under pom.xml. Then I need to create an instance of that caching service and *tell* that caching service to use the Redis service I entered in the application.properties file.
3. Similar to the WIP code I have, [this](https://github.com/dewan-ahmed/quarkus-redis-cache/blob/1e6e0d5f1adfb8d79595c5a1b68e4df25a5c8826/src/main/java/org/acme/cache/GithubStarsService.java#L23) is the only place where I'll need to *swap* the caching mechanism.

I tried to follow [this guide](https://redisson.org/articles/redis-integration-with-quarkus.html) but couldn't make much progress. Please let me know any guide/resource I can follow and if my above assumptions are correct. Thanks.

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
Instead of redisson, you should use the quarkus-redis-client. 

However, I'm not sure that our cache layer is "pluggable". That's a question for @karesti I think.

**Reply by gwenneg** [reply]
`quarkus-cache` was designed to support multiple caching providers, but local caching with Caffeine is the only available option for now. Using another provider would require changes in the extension code so it's not pluggable indeed.

**Reply by machi1990** [reply]
@gwenneg I think this could be a very useful feature for both extensions (redis and cache).

**Reply by dewan-ahmed** [reply]
Thanks for your responses. Is there a guide I can follow that shows the usage of remote cache (using `quarkus-redis-client`) instead of the local cache?

**Reply by dewan-ahmed** [reply]
> `quarkus-cache` was designed to support multiple caching providers, but local caching with Caffeine is the only available option for now. Using another provider would require changes in the extension code so it's not pluggable indeed.

Hi @gwenneg could you please point me to a doc/guide on how I can use `quarkus-cache` with an external Redis caching provider? I tried to search the Quarkus docs but couldn't find much details.

**Reply by karesti** [reply]
I'm working on having the annotations in the infinispan client extension ATM :)

**Reply by dewan-ahmed** [reply]
That's great @karesti 

When implemented, would the infinispan client use the underlying **Caffeine cache + Infinispan distributed cache**, Or, the Quarkus app will directly use the Infinispan distributed cache?

**Reply by gwenneg** [reply]
As I wrote earlier, the `quarkus-cache` extension only supports local caching with Caffeine right now. From a Quarkus user perspective, there is no way to use Redis as the caching provider with that extension. This could change in a future Quarkus release though if someone contributes to the extension and adds a Redis caching provider.

`quarkus-infinispan-client` which was mentioned by @karesti is a different extension that offers distributed caching. It is based on Infinispan and does not have anything to do with Redis AFAIK, but I'm not an expert about this extension.

If you really want to use Redis, then `quarkus-redis-client` may be the only available solution right now. It won't work with the annotations from `quarkus-cache`.

**Reply by gsmet** [reply]
There is now a `quarkus-redis-cache` extension that does exactly what you wanted: https://quarkus.io/guides/cache-redis-reference .
